provider/custom_server.go

```diff
@@ -0,0 +1,306 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+
+	"github.com/go-pkgz/auth/avatar"
+	"github.com/go-pkgz/auth/logger"
+	"github.com/go-pkgz/auth/token"
+	"golang.org/x/oauth2"
+	"gopkg.in/oauth2.v3/server"
+)
+
+const custOAuthPort = 9096
+
+// CustomProviderOpt are options to start go-oauth2/oauth2 server
+type CustomProviderOpt struct {
+	WithLoginPage bool
+	Cid           string
+}
+
+// CustomOauthServer is a go-oauth2/oauth2 server running on its own port
+type CustomOauthServer struct {
+	logger.L
+	// Domain corresponds to the root host specified without port
+	Domain string
+	// WithLoginPage: redirect to login html page if true
+	WithLoginPage bool
+	httpServer    *http.Server
+	// OauthServer is instance of go-oauth2/oauth2 server
+	OauthServer *server.Server
+	lock        sync.Mutex
+}
+
+// Run starts serving on custOauthPort
+func (c *CustomOauthServer) Run(ctx context.Context) {
+	c.Logf("[INFO] run local go-oauth2/oauth2 server on %s:%d", c.Domain, custOAuthPort)
+	c.lock.Lock()
+
+	c.httpServer = &http.Server{
+		Addr: fmt.Sprintf(":%d", custOAuthPort),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			switch {
+			case strings.HasSuffix(r.URL.Path, "/authorize"):
+				c.handleAuthorize(w, r)
+			case strings.HasSuffix(r.URL.Path, "/access_token"):
+				if err := c.OauthServer.HandleTokenRequest(w, r); err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+				}
+			case strings.HasPrefix(r.URL.Path, "/user"):
+				c.handleUserInfo(w, r)
+			case strings.HasPrefix(r.URL.Path, "/avatar"):
+				c.handleAvatar(w, r)
+			default:
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+		}),
+	}
+	c.lock.Unlock()
+
+	go func() {
+		<-ctx.Done()
+		c.Logf("[DEBUG] cancellation via context, %v", ctx.Err())
+		c.Shutdown()
+	}()
+
+	err := c.httpServer.ListenAndServe()
+	c.Logf("[WARN] dev oauth2 server terminated, %s", err)
+}
+
+func (c *CustomOauthServer) handleAuthorize(w http.ResponseWriter, r *http.Request) {
+	// Called for first time, ask for username
+	if c.WithLoginPage && (r.ParseForm() != nil || r.Form.Get("username") == "") {
+		userLoginTmpl, err := template.New("page").Parse(customLoginTmpl)
```

Comment:
Test coverage for this file is below 70% and this branch of code (within `if c.WithLoginPage && ...`) not covered at all.

_Not sure if you can see coveralls report, but in case if it open publicly see https://coveralls.io/builds/24836939/source?filename=provider%2Fcustom_server.go#L80_

Reply:
Yes, I can see coveralls report. I had to check test coverage before submitting pull request. I'll take care of it.